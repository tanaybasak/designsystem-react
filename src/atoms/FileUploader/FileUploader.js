import React from 'react';
import PropTypes from 'prop-types';
import prefix from "../../settings";

export default function FileUploader({ id, className, label, description, disabled, multiple, tabIndex, onChange, ...restProps }) {
    const classnames = `${prefix}-file-btn ${className}`.trim();

    return (
        <div className={`${prefix}-form-item`}>
            <span className={`${prefix}-label`}>{label}
            </span>
            <p className={`${prefix}-label-description`}>{description}</p>
            <div className={`${prefix}-file-uploader`} {...restProps}>
                <input type="file" className={`${prefix}-file-input`} id={id} tabIndex={tabIndex}
                    onChange={onChange} disabled={disabled} multiple={multiple} />
                <label for={id} className={classnames} role="button">
                    Add file
                </label>
            </div>
        </div>
    );
};

FileUploader.propTypes = {
    id: PropTypes.string.isRequired,
    className: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    multiple: PropTypes.bool,
    tabIndex: PropTypes.number,
    onChange: PropTypes.func
};

FileUploader.defaultProps = {
    id: null,
    className: '',
    label: '',
    description: '',
    disabled: false,
    multiple: true,
    tabIndex: 0,
    onChange: () => { }
};