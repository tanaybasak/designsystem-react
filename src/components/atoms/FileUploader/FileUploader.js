import React from 'react';
import PropTypes from 'prop-types';
import { prefix } from '../../../settings';

export default function FileUploader({ id, className, label, description, disabled, multiple, tabIndex, onChange, ...restProps }) {
    const classnames = `${prefix}-file ${className}`.trim();

    return (
        <div className={`${prefix}-form-item`}>
            <strong className={`${prefix}-label`}>{label}
            </strong>
            <p className={`${prefix}-label-description`}>{description}</p>
            <div className={classnames} {...restProps}>
                <input type="file" className={`${prefix}-file-input`} id={id} tabIndex={tabIndex}
                    onChange={onChange} disabled={disabled} multiple={multiple} />
                <label for={id} className={`${prefix}-file-btn ${prefix}-btn ${prefix}-primary
                    ${prefix}-primary-outline ${prefix}-sm`} role="button" tabindex="0">
                    Add file
                    <span className={`${prefix}-file-btn-icon`}></span>
                </label>
            </div>
        </div>
    );
};

FileUploader.propTypes = {
    id: PropTypes.string.isRequired,
    className: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    multiple: PropTypes.bool,
    tabIndex: PropTypes.number,
    onChange: PropTypes.func
};

FileUploader.defaultProps = {
    id: null,
    className: '',
    label: '',
    description: '',
    disabled: false,
    multiple: true,
    tabIndex: 0,
    onChange: () => { }
};