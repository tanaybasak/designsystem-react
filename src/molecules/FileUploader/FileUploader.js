import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import prefix from '../../settings';
import Paragraph from '../../atoms/Paragraph/Paragraph';

export default function FileUploader({
  id,
  className,
  label,
  children,
  description,
  disabled,
  multiple,
  fileType,
  tabIndex,
  hideFile,
  onChange,
  ...restProps
}) {
  const classnames = `${prefix}-file-btn ${prefix}-btn ${className}`.trim();

  const [fileLists, setFileList] = useState([]);

  const fileContainer = useRef(null);

  const keyListener = event => {
    if (event.key === 'Enter') {
      event.preventDefault();
      fileContainer.current.querySelector(`.${prefix}-file-btn`).click();
    }
  };

  const getFileList = event => {
    const files = event.target.files;
    const filelist = Object.keys(files).map(i => files[i]);
    const tempFileLists = multiple
      ? [...fileLists, ...filelist]
      : [...filelist];
    setFileList(tempFileLists);
    onChange(tempFileLists, event);
    //event.target.value = null;
  };

  const removeFile = (event, name) => {
    event.preventDefault();
    const index = fileLists.findIndex(file => file.name === name);
    const newFileList = [...fileLists];
    if (index !== -1) {
      newFileList.splice(index, 1);
      setFileList(newFileList);
    }
    onChange(newFileList, event);
  };

  return (
    <div className={`${prefix}-form-item`}>
      {label && label.length ? (
        <span className={`${prefix}-label`}>{label}</span>
      ) : null}
      {description && description.length ? (
        <Paragraph className={`${prefix}-label-description`}>
          {description}
        </Paragraph>
      ) : null}
      <div
        className={`${prefix}-file-uploader`}
        ref={fileContainer}
        {...restProps}
      >
        <input
          type="file"
          className={`${prefix}-file-input`}
          id={id}
          tabIndex={tabIndex}
          onChange={getFileList}
          disabled={disabled}
          multiple={multiple}
          accept={fileType}
        />
        <label
          htmlFor={id}
          className={classnames}
          onKeyDown={keyListener}
          tabIndex={!disabled ? '0' : null}
        >
          {children}
        </label>
        <div className={`${prefix}-file-container`}>
          {fileLists.length && !hideFile
            ? fileLists.map((fileList, index) => (
                <div key={index} className={`${prefix}-file-container-item`}>
                  <span
                    title={fileList.name}
                    className={`${prefix}-file-selected-file`}
                  >
                    <p className={`${prefix}-file-filename`}>{fileList.name}</p>
                  </span>
                  <button
                    aria-label="close file"
                    onClick={e => removeFile(e, fileList.name)}
                    type="button"
                    className={`${prefix}-file-close`}
                  />
                </div>
              ))
            : null}
        </div>
      </div>
    </div>
  );
}

FileUploader.propTypes = {
  /** Unique id for File Uploader */
  id: PropTypes.string.isRequired,
  /** Name of the custom class to apply to the File Uploader Button
   * eg:
   * Primary: 'hcl-btn hcl-primary',
   * Primary Danger: 'hcl-btn hcl-primary hcl-danger',
   * Primary Danger Small: 'hcl-btn hcl-primary hcl-danger hcl-sm',
   * Primary Small: 'hcl-btn hcl-primary hcl-sm',
   * Secondary: 'hcl-btn hcl-secondary',
   * Secondary Danger: 'hcl-btn hcl-secondary hcl-danger',
   * Secondary Danger Small: 'hcl-btn hcl-secondary hcl-danger hcl-sm',
   * Ghost: 'hcl-btn hcl-ghost'
   */
  className: PropTypes.string,
  /** Label for File Uploader */
  label: PropTypes.string,
  /** Children of File Uploader button */
  children: PropTypes.any,
  /** Description of File Uploader */
  description: PropTypes.string,
  /** Boolean value to disable File Uploader */
  disabled: PropTypes.bool,
  /** Boolean value to select Multiple files from File Uploader */
  multiple: PropTypes.bool,
  /** File type for File Uploader */
  fileType: PropTypes.string,
  /** Tab Index for File Uploader */
  tabIndex: PropTypes.number,
  /** Call back function that is invoked when File Uploader is clicked */
  onChange: PropTypes.func,
  /** Boolean value to hide or show file names selected from File Uploader */
  hideFile: PropTypes.bool
};

FileUploader.defaultProps = {
  id: null,
  className: '',
  label: '',
  children: null,
  description: '',
  disabled: false,
  multiple: false,
  fileType: '',
  tabIndex: 0,
  hideFile: false,
  onChange: () => {}
};
