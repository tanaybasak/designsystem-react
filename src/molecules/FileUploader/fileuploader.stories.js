import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { select, text, boolean } from '@storybook/addon-knobs';
//@update-path-build-start
import FileUploader from './FileUploader';
//@update-path-build-end
import '../../story.css';
import 'patron-css/dist/patron-style.css';

const classOptions = {
  Primary: 'hcl-btn hcl-primary',
  'Primary Danger': 'hcl-btn hcl-primary hcl-danger',
  'Primary Danger Small': 'hcl-btn hcl-primary hcl-danger hcl-sm',
  'Primary Small': 'hcl-btn hcl-primary hcl-sm',
  Secondary: 'hcl-btn hcl-secondary',
  'Secondary Danger': 'hcl-btn hcl-secondary hcl-danger',
  'Secondary Danger Small': 'hcl-btn hcl-secondary hcl-danger hcl-sm',
  Ghost: 'hcl-btn hcl-ghost'
};

storiesOf('FileUploader', module).add(
  'basic',
  () => (
    <FileUploader
      className={select(
        'Button Type',
        classOptions,
        'hcl-btn hcl-primary hcl-sm'
      )}
      description={text('Description', 'File Uploader Description')}
      disabled={boolean('Disabled', false)}
      fileType={text('File Type', '')}
      id="sample_file_uploader"
      label={text('Label', 'Add File')}
      multiple
      onChange={action('File Uploader OnChange')}
    >
      Add File
    </FileUploader>
  ),
  {
    info: {
      text: 'Description About FileUploader Component'
    }
  }
);
