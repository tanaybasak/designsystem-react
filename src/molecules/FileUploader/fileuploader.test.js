import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import prefix from '../../settings';
import FileUploader from './FileUploader';

describe('<FileUploader> component', () => {

  it('snapshots/renders Button correctly', () => {
    const tree = renderer
      .create(<FileUploader id="file_uploader">Add File</FileUploader>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with expected id', () => {
    const mockCallBack = jest.fn();
    const fileUploader = mount(
      <FileUploader id="file_uploader" onClick={mockCallBack}>
        Add File
      </FileUploader>
    );
    expect(fileUploader.find(`#file_uploader`).exists()).toBeTruthy();
  });

  it('renders with button className', () => {
    const mockCallBack = jest.fn();
    const fileUploader = mount(
      <FileUploader id="file_uploader" className="hcl-secondary" onClick={mockCallBack}>
        Add File
      </FileUploader>
    );
    expect(fileUploader.exists(`.${prefix}-secondary`)).toEqual(true);
  });

  it('renders with default multiple prop', () => {
    const mockCallBack = jest.fn();
    const fileUploader = mount(
        <FileUploader id="file_uploader" multiple onClick={mockCallBack}>
          Add File
        </FileUploader>
      );
    expect(fileUploader.props().multiple).toEqual(true);
  });

  it('add files', () => {
    const mockCallBack = jest.fn();
    const file = new File(['dummy content'], 'example.png', {type: 'image/png'})

    const mockedEvent = { target: { files: [file] }}

    const fileUploader = mount(
        <FileUploader id="file_uploader" multiple onClick={mockCallBack}>
          Add File
        </FileUploader>
      ); 
      fileUploader.find(`.${prefix}-file-input`).simulate('change',mockedEvent);
     expect(fileUploader.exists(`.${prefix}-file-filename`)).toEqual(true);
     expect(fileUploader.find(`.${prefix}-file-filename`).text()).toEqual('example.png');
  });

  it('remove files', () => {
    const mockCallBack = jest.fn();
    const file = new File(['dummy content'], 'example.png', {type: 'image/png'})

    const mockedEvent = { target: { files: [file] }}

    const fileUploader = mount(
        <FileUploader id="file_uploader" multiple onClick={mockCallBack}>
          Add File
        </FileUploader>
      );
    fileUploader.find(`.${prefix}-file-input`).simulate('change',mockedEvent); 
    fileUploader.find(`.${prefix}-file-close`).simulate('click',{name : 'example.png'});
     expect(fileUploader.exists(`.${prefix}-file-filename`)).toEqual(false);
  });
});
