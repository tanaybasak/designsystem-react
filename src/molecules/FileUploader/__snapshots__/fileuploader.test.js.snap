// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileUploader> component snapshots/renders Button correctly 1`] = `
<div
  className="hcl-form-item"
>
  <div
    className="hcl-file-uploader"
  >
    <input
      accept=""
      className="hcl-file-input"
      disabled={false}
      id="file_uploader"
      multiple={false}
      name=""
      onChange={[Function]}
      required={false}
      tabIndex={0}
      type="file"
    />
    <label
      className="hcl-file-btn hcl-btn"
      htmlFor="file_uploader"
      onKeyDown={[Function]}
      tabIndex="0"
    >
      Add File
    </label>
    <div
      className="hcl-file-container"
    />
  </div>
</div>
`;
